### ISSUE TITLE
Qt 6.2.x issues

### ISSUE THREAD

**Issue by dae** [ankitects/anki#1427]
Tracking issues in Qt6.2 that did not exist in our 2.1.49 release.

macOS:

- [x] the Esc key is swallowed in input elements (fix in 6.2.1: https://bugreports.qt.io/browse/QTBUG-97119)
- [x] recording crashes on darwin-arm64 (https://bugreports.qt.io/browse/QTBUG-97838)
- [ ] not new in Qt6, but we may want to pass --disable-gpu automatically when set to software mode on macOS (https://forums.ankiweb.net/t/anki-opens-but-all-windows-are-blank/8920/18)
- [ ] ~cmd+option+shift+c for 'same cloze' is broken: https://forums.ankiweb.net/t/anki-2-1-50-beta/15608/68?u=dae~ split into https://github.com/ankitects/anki/issues/1748
- [ ] ~tabbed layout not possible (from Qt 5.15 on: https://bugreports.qt.io/browse/QTBUG-85907)~ split into https://github.com/ankitects/anki/issues/1752
- [ ] ~IME gets confused when changing windows: https://forums.ankiweb.net/t/anki-2-1-50-beta-6-7-stable-release/18181/39~ split into https://github.com/ankitects/anki/issues/1750

Windows:

- [x] alt-gr conflicting with shortcuts: https://forums.ankiweb.net/t/altgr-issues-on-anki-windows-altgr-4-for-and-altgr-2-for-do-not-work/14588/7
- [ ] ~6.2.3 inserts junk into the clipboard; we won't be able to upgrade past 6.2.2 until https://bugreports.qt.io/browse/QTBUG-100806 is fixed~ split into https://github.com/ankitects/anki/issues/1751

All platforms:
- [ ] ~large images load slowly (started in 5.15, worse in 6.2; fix likely in webengine 6.2.4) (https://bugreports.qt.io/browse/QTBUG-99130)~ split into https://github.com/ankitects/anki/issues/1753

Previous ticket about Qt 5.15 issues: #982

**Comment by hgiesel** [comment]
Since Qt6.2 I often get this issue when reopening the editor after having had Anki run in the background for some time:

> [1020/221032.469448:ERROR:icu_util.cc(252)] Couldn't mmap icu data file

It causes the program to hang, and I have to shut it down via the task manager. Don't know if that only happens when running from source or also in the builds.

**Comment by RumovZ** [comment]
> alt-gr conflicting with shortcuts

This may not belong here. I successfully reproduced the bug in Qt 6.2.2 outside Anki. However, I can also reproduce it in Anki 2.1.50-qt5. But doesn't that use the same Qt release as 2.1.49, in which the bug is not reproducible? This would mean there are two separate bugs in Qt and PyQt. Maybe someone can verify my findings.
As a sidenote, the Pyqt6 designer contains this bug as well.

**Comment by dae** [comment]
2.1.49 uses PyQt 5.14 on Windows. 2.1.50 needed to bump that to 5.15, as the 5.14 wheels don't support Python 3.9 on Windows. So it sounds like this may be a change in behaviour in 5.15 that has persisted into Qt 6.2.

I think this may boil down to an issue I previously reported to them 😅 : https://bugreports.qt.io/browse/QTBUG-73247. If you have the Qt 6.2 install handy and can reproduce it, hopefully posting back there to let them know will encourage them to take another look. It might be worth testing both webview and non-webview cases - a number of issues in the past have happened when inputting into a webview, but didn't happen when inputing into something like a QLineEdit, as there seem to be quite a few rough edges in the way they integrated Chromium into the rest of Qt.

Possibly related: in `aqt/__init__.py` we have this code:

```python
    if (
        is_win
        and qtmajor == 5
        and (qtminor == 14 or (qtminor == 15 and qtpoint == 0))
        and "QT_QPA_PLATFORM" not in os.environ
    ):
        os.environ["QT_QPA_PLATFORM"] = "windows:altgr"
```

I think they may have have made it the default in subsequent releases, but it might be worth giving it a try to see if it changes anything.

**Comment by RumovZ** [comment]
> 2.1.49 uses PyQt 5.14 on Windows.

Sorry, reading is hard...

> It might be worth testing both webview and non-webview cases - a number of issues in the past have happened when inputting into a webview, but didn't happen when inputing into something like a QLineEdit, as there seem to be quite a few rough edges in the way they integrated Chromium into the rest of Qt.

I've always tested it with a QLineEdit, so I guess the webview is not to blame this time.

> I think this may boil down to an issue I previously reported to them 😅 : https://bugreports.qt.io/browse/QTBUG-73247. If you have the Qt 6.2 install handy and can reproduce it, hopefully posting back there to let them know will encourage them to take another look. 

Done.

> I think they may have have made it the default in subsequent releases, but it might be worth giving it a try to see if it changes anything.

What am I supposed to do with this code? Is it supposed to fix or break things? 😅

**Comment by dae** [comment]
The 5.x version remaining the same would be the logical thing, so I don't blame you :-)

Thanks for doing that testing. Now we cross our fingers and hope someone actually looks into it!

I'm unsure exactly what the setting does, I only recall it was required in the past, and I was under the impression the fix was integrated into Qt - just thought it might be worth trying to set it on an updated version to confirm it doesn't change anything.

**Comment by RumovZ** [comment]
> I'm unsure exactly what the setting does, I only recall it was required in the past, and I was under the impression the fix was integrated into Qt - just thought it might be worth trying to set it on an updated version to confirm it doesn't change anything.

It works! 🥳 So this explains the breakage with 2.1.50-qt5. There never was a fix for 5.14.
And sorry, I could have figured out sooner what to do with that snippet, if I had looked properly. I'll fix my report over on https://bugreports.qt.io.

**Comment by dae** [comment]
Nice easy win 🎉 

Just a gentle nudge about that bug report. If you're feeling up for it, it would be a big help if you could also file a report about the image loading issue as well (presumably it can be reproduced with one of the Qt browser examples): https://forums.ankiweb.net/t/anki-2-1-50-beta/15608/55

cqg's large file can be grabbed here from https://apps.ankiweb.net/downloads/large.png, and then a file:/// url could probably be used in the example to avoid network speed factoring into it.

**Comment by RumovZ** [comment]
Reported here: https://bugreports.qt.io/browse/QTBUG-99130

**Comment by dae** [comment]
Great, thanks for that Rumo - I've also linked to it on https://github.com/ankitects/anki/issues/1427

**Comment by hgiesel** [comment]
Another shortcut sounds good. Now that we only enable the shortcut on cloze cards (we could enable it on only on cloze fields even), maybe we can change the shortcuts to Cmd+D and Cmd+Shift+D, and sell it as a feature (easier shortcut) :)

**Comment by dae** [comment]
Unfortunately cmd+d triggers the deck selector, so we'd probably need to keep the standard cloze shortcut the same. 

I did a little experimenting and it seems like if we use keyup instead of keydown, the shortcut does work correctly. Would it cause other issues if we switched to that instead? Because this issue is a little niche, I'm not very confident Qt will be in a rush to fix it, so if we can find a workaround on our end that will likely unblock us faster :-)
